Improve documentation and error messages for safe initialization (#15659)
lampepfl/dotty: docs/_docs/reference/other-new-features/safe-initialization.md

   `List.apply(e)`, the argument `e` may be non-hot. If that is the case, the
   result value of the parametric method call is taken as _cold_.

5. Method calls on hot values with effectively hot arguments produce hot results.

   This rule is assured by local reasoning about initialization.

6. Method calls on `ThisRef` and warm values will be resolved statically and the
   corresponding method bodies are checked.

7. In a new expression `new p.C(args)`, if the values of `p` and `args` are
   effectively hot, then the result value is also hot.

   This is assured by local reasoning about initialization.

8. In a new expression `new p.C(args)`, if any value of `p` and `args` is not
   effectively hot, then the result value takes the form `Warm[C] { outer = Vp, args = Vargs }`. The initialization code for the class `C` is checked again to make
   sure the non-hot values are used properly.

   In the above, `Vp` is the widened value of `p` --- the widening happens if `p`
   is a warm value `Warm[D] { outer = V, args }` and we widen it to
   `Warm[D] { outer = Cold, args }`.

   The variable `Vargs` represents values of `args` with non-hot values widened
   to `Cold`.

>    The widening is motivated to finitize the abstract domain and ensure

Reviewer: ```suggestion
   The motivation for the widening is to finitize the abstract domain and ensure
```

>    terimination of the initialization check.

Reviewer: ```suggestion
   termination of the initialization check.
```

> 9. The scrutinee in pattern match, the values in return and throw statements must be _effectively hot_.

Reviewer: ```suggestion
9. The scrutinee in a pattern match and the values in return and throw statements must be _effectively hot_.
```

> A value `v` is _effectively hot_ if any of the following is true:

Reviewer: Move this definition before "Rules".

- `v` is `Hot`.

> - `v` is `ThisRef` and all fields of the underlying object are initialized.

Reviewer: ```suggestion
- `v` is `ThisRef` and all fields of the underlying object are assigned.
```

- `v` is `Warm[C] { outer = V, ctor, args = Vs }` and
  1. `C` does not contain inner classes;

>   2. Calling any method on `v` encounters no errors and the method return value is _effectively hot_;

Reviewer: ```suggestion
  2. Calling any method on `v` encounters no initialization errors and the method return value is _effectively hot_;
```

>   2. Calling any method on `v` encounters no errors and the method return value is _effectively hot_;

Reviewer: ```suggestion
  2. Calling any method on `v` encounters no errors and the method return value is _effectively hot_; and
```

  3. Each field of `v` is _effectively hot_.
- `v` is `Fun(e, V, C)` and calling the function encounters no errors and the
  function return value is _effectively hot_.
- `ThisRef` is _effectively hot_.

> 

Reviewer: ```suggestion
In the `ThisRef` case, local reasoning ensures that all fields assigned implies that all fields themselves contain effectively hot values.

In the `Warm` case, if the value `v` leaks outside the analysis scope, the unanalyzed code could instantiate inner classes of `v`, call methods on `v`, or read fields of `v`. The conditions of the `Warm` case ensure that these actions do not lead to accessing a cold object.

In the `Fun` case, ...
```

Author: Maybe we can add the justification later --- they are not immediately understandable to readers and can itself be an obstacle in reading the docs.
